cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,204 @@
+    return index == 0 || not comp(perm[index], perm[index - 1]);
+  };
+
+private:
+  row_equality_comparator<has_nulls> comp;
+  size_type const *perm;
+};
+
+std::unique_ptr<table> rank(table_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const size = input.num_rows();
+
+  std::vector<std::unique_ptr<column>> rank_columns;
+  for (auto const &input_col : input) {
+    std::unique_ptr<column> sorted_order =
+        (method == rank_method::FIRST)
+            ? detail::stable_sorted_order(table_view{{input_col}},
+                                          {column_order}, {null_precedence}, mr,
+                                          stream)
+            : detail::sorted_order(table_view{{input_col}}, {column_order},
+                                   {null_precedence}, mr, stream);
+    column_view sorted_order_view = sorted_order->view();
+
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, copy_bitmask(input_col, stream, mr),
+          input_col.null_count(), stream, mr));
+    else
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, mask_state::UNALLOCATED, stream, mr));
+
+    auto rank_mutable_view = rank_columns.back()->mutable_view();
+    auto rank_data = rank_mutable_view.data<double>();
+    auto device_table =
+        table_device_view::create(table_view{{input_col}}, stream);
+    rmm::device_vector<double> dense_rank_sorted; // as key for min, max,
+                                                  // average
+    // TODO: double? or size_type?
+    if (method != rank_method::FIRST) {
+      dense_rank_sorted = rmm::device_vector<double>(size);
+      if (input_col.has_nulls()) {
+        auto conv = unique_comparator<true, double>(
+            *device_table, sorted_order_view.data<size_type>());
+        auto it = thrust::make_transform_iterator(
+            thrust::make_counting_iterator<size_type>(0), conv);
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), it,
+                               it + size,
+                               dense_rank_sorted.data().get());
+      } else {
+        auto conv = unique_comparator<false, double>(
+            *device_table, sorted_order_view.data<size_type>());
+        auto it = thrust::make_transform_iterator(
+            thrust::make_counting_iterator<size_type>(0), conv);
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), it,
+                               it + size,
+                               dense_rank_sorted.data().get());
+      }
+    }
+
+    switch (method) {
```

Comment:
Please document each case so the maintainer doesn't have to refer to the pandas docs to understand what is happening here.

Reply:
added a short description of algorithm and method.